### PR TITLE
Peephole log2(const - x) to log2(max(0.0, const - x)).

### PR DESCRIPTION
**Pull request by perlfu**
This avoids edge case precision underflows in poorly defined applications which are insufficiently tolerant to Vulkan math reordering rules.

### CODE REVIEW

**Comment by amdvlk-admin** [comment]
### Test summary for commit e64a1357971cdb7bb43c5a64a33db79ed227f300
<details><summary>CTS tests (Failed: 0/142036)</summary><ul><li>Built with version 1.3.3.1</li>
<details open><summary>Ubuntu 22.04, Navi3x</summary><ul>

<li>Passed: 40426/71016 (56.9%)</li>
<li>Failed: 0/71016 (0.0%)</li>
<li>Not Supported: 30590/71016 (43.1%)</li><li>Warnings: 0/71016 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Navi2x</summary><ul>

<li>Passed: 40464/71020 (57.0%)</li>
<li>Failed: 0/71020 (0.0%)</li>
<li>Not Supported: 30556/71020 (43.0%)</li><li>Warnings: 0/71020 (0.0%)</li>
</ul></details>


</ul></details>

File: lgc/patch/PatchPeepholeOpt.cpp [hunk 1]
```diff
@@ -143,4 +146,37 @@ void PatchPeepholeOpt::visitIntToPtr(IntToPtrInst &intToPtr) {
...
+void PatchPeepholeOpt::visitCallInst(CallInst &callInst) {
+  if (callInst.getIntrinsicID() != Intrinsic::log2)
+    return;
+
+  Value *src = callInst.getOperand(0);
+  Instruction *srcInst = dyn_cast<Instruction>(src);
+  if (!srcInst || srcInst->getOpcode() != Instruction::FSub)
+    return;
+
+  // Do not touch instructions marked as potentially having NaNs, as this would not be a legal optimisation.
```
Comment (nhaehnle) [inline lgc/patch/PatchPeepholeOpt.cpp:165]: Optimisation is the wrong word here. Call it a transformation.

File: lgc/test/PeepholeOptLog2Underflow.lgc [hunk 2]
```diff
@@ -0,0 +1,27 @@
+; Test that log2 underflow optimisation is correctly applied.
```
Comment (nhaehnle) [inline lgc/test/PeepholeOptLog2Underflow.lgc:1]: Again, not an optimisation...

**Review by nhaehnle — COMMENTED** [review]
This should probably apply to `x - c` and `x + c` / `c + x` as well. I'd suggest using the PatternMatch.h infrastructure for that.

**Comment by perlfu** [comment]
- Rebase
- Remove unintended use of word "optimisation"
- Use PatternMatch infrastructure
- Extend pattern to FAdd, and reordering of constant and FAdd/FSub
- Extend test to cover above

**Comment by amdvlk-admin** [comment]
### Test summary for commit 9b464d4e08cbb763af93c0791b29dfed39aba5bb
<details><summary>CTS tests (Failed: 0/142050)</summary><ul><li>Built with version 1.3.3.1</li>
<details open><summary>Ubuntu 22.04, Navi3x</summary><ul>

<li>Passed: 40426/71030 (56.9%)</li>
<li>Failed: 0/71030 (0.0%)</li>
<li>Not Supported: 30604/71030 (43.1%)</li><li>Warnings: 0/71030 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Navi2x</summary><ul>

<li>Passed: 40464/71020 (57.0%)</li>
<li>Failed: 0/71020 (0.0%)</li>
<li>Not Supported: 30556/71020 (43.0%)</li><li>Warnings: 0/71020 (0.0%)</li>
</ul></details>


</ul></details>

File: lgc/test/PeepholeOptLog2Underflow.lgc [hunk 3]
```diff
@@ -0,0 +1,47 @@
+; Test that log2 underflow transform is correctly applied.
+
+; RUN: lgc -mcpu=gfx1010 -print-after=lgc-patch-peephole-opt -o - - <%s 2>&1 | FileCheck --check-prefixes=CHECK %s
```
Comment (nhaehnle) [inline lgc/test/PeepholeOptLog2Underflow.lgc:3]: I think this test can just use the -passes=lgc-patch-peephole-opt and update_test_checks. And moving it to lgc/test/Transforms/PeepholeOpt for consistency would be grand.

**Review by nhaehnle — COMMENTED** [review]
Thanks. Just one nit about the test, the rest looks good to me.

**Comment by perlfu** [comment]
- Rebase
- Rename test (into subdirectory)
- Generate test using update_test_checks

**Comment by amdvlk-admin** [comment]
### Test summary for commit 23e338422855ef90c32d1641a4ca453f9718233d
<details><summary>CTS tests (Failed: 0/141320)</summary><ul><li>Built with version 1.3.3.1</li>
<details open><summary>Ubuntu 22.04, Navi3x</summary><ul>

<li>Passed: 40065/70667 (56.7%)</li>
<li>Failed: 0/70667 (0.0%)</li>
<li>Not Supported: 30602/70667 (43.3%)</li><li>Warnings: 0/70667 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Navi2x</summary><ul>

<li>Passed: 40106/70653 (56.8%)</li>
<li>Failed: 0/70653 (0.0%)</li>
<li>Not Supported: 30547/70653 (43.2%)</li><li>Warnings: 0/70653 (0.0%)</li>
</ul></details>


</ul></details>

**Review by nhaehnle — APPROVED** [review]
Thanks!

File: lgc/patch/PatchPeepholeOpt.cpp [hunk 4]
```diff
@@ -143,4 +148,35 @@ void PatchPeepholeOpt::visitIntToPtr(IntToPtrInst &intToPtr) {
...
+// This addresses a potential precision underflow in applications intolerant to in-spec math reordering.
+//
+// @param callInst: The call instruction to visit.
+void PatchPeepholeOpt::visitCallInst(CallInst &callInst) {
+  if (callInst.getIntrinsicID() != Intrinsic::log2)
+    return;
+
+  Value *V = callInst.getOperand(0);
+
+  if (!(match(V, m_FSub(m_Constant(), m_Value())) || match(V, m_FSub(m_Value(), m_Constant())) ||
```
Comment (jayfoad) [inline lgc/patch/PatchPeepholeOpt.cpp:164]: FYI there is m_c_BinOp(opcode, lhs, rhs) for matching with operands in either order.
